src/slurm_plugin/instance_manager.py

```diff
@@ -358,3 +383,43 @@ def terminate_all_compute_nodes(self, terminate_batch_size):
     def _update_failed_nodes(self, nodeset, error_code="Exception"):
         """Update failed nodes dict with error code as key and nodeset value."""
         self.failed_nodes[error_code] = self.failed_nodes.get(error_code, set()).union(nodeset)
+
+    def get_node_instance_mapper(self):
+        """Return a function that maps nodes to instance ids."""
+
+        def get_instances_for_nodes(nodes):
+            nodes_with_instance_id = []
+            nodes_without_instance_id = []
+            for node in nodes:
+                (nodes_with_instance_id if node.instance else nodes_without_instance_id).append(
+                    {
+                        "Name": node.name,
+                        "InstanceId": node.instance.id if node.instance else None,
+                    }
+                )
+
+            # Return instance ids that don't require a DDB lookup first.
+            for node in nodes_with_instance_id:
+                yield node
+
+            # Only get a ddb resource if we have something to do.
+            if len(nodes_without_instance_id) < 1:
+                return
+
+            # Lookup instance ids in DDB for nodes that we don't already have the instance id for.
+            ddb_resource = resource_factory("dynamodb")
```

Comment:
Interesting - are partials preferable to using closures in Python?

This code was refactored so that the factory and table name are passed in as parameters to the enclosing method. lmkwyt...